File: Day1-Frequency.cpp
#include <iostream>
#include <fstream>
#include <list>
#include <algorithm>
using namespace std;

int main(){
    int sum = 0;
    int change = 0;
    list<int> changeList;
    list<int> freqList;
    list<int>::iterator dupFreq, freqChange;
    ifstream file;
    bool isDup = false;

    freqList.push_front(0); // 0 frequency is always on the list.
    file.open("Day1-input.txt");
    if(!file){ //Check if the file is successfully opened.
        cout << "Cannot open input file.";
        return(1);
    }

    //First, read all the freq changes into the list
    while(file >> change){
        changeList.push_back(change);
    }
    file.close();

    while(!isDup){
        freqChange = changeList.begin();
        while(freqChange != changeList.end()){
            sum += *freqChange; // Change the frequency
            freqChange++;

            //Try to find if the most recent frequency is duplicated
            dupFreq = find(freqList.begin(), freqList.end(), sum);
            if(dupFreq != freqList.end()){ //If it's duplicated
                cout << "Freq " << sum << " is duplicated" << endl;
                isDup = true;
                break;
            } else{
                //If it's the unique frequency, remember the number
                freqList.push_back(sum);
            }

        }
    }
    return 0;
}

File: Day1-Frequency.py
def check_frequency(filename: str):
    input_file = open(filename)
    freq_change_list = input_file.readlines()
    freq_list = set()  # Use set instead of list because of element searching performance
    frequency = 0
    is_duplicated = False

    round = 0
    while not is_duplicated:  # Perform frequency changing until the duplicate number has been found
        for freq_change in freq_change_list:
            frequency += int(freq_change)  # Change the frequency

            if frequency in freq_list:  # Check if this frequency has already existed
                print("Freq ", frequency, " is duplicated")
                is_duplicated = True  # Set the duplication flag
                break
            else:
                freq_list.add(frequency)  # If it's not duplicated, add to the frequency set


check_frequency("Day1-input.txt")
